_inc/client/components/number-format/index.js

```diff
@@ -0,0 +1,54 @@
+/**
+ * WordPress dependencies
+ */
+import { __experimentalGetSettings } from '@wordpress/date';
+
+/**
+ * Clean up WP locale so it matches the format expected by browsers.
+ *
+ * @param {string} locale - Locale given by WordPress.
+ *
+ * @returns {string} Browser-formatted locale.
+ */
+export const cleanLocale = locale => {
+	return (
+		locale
+			// Keep only the basic locale (WP locales can have variants, such as de_DE_formal).
+			.replace( /(\w{2})?(_)?(\w{2})?(?:_)?(?:\w*)?/, '$1$2$3' )
```

Comment:
Can WordPress locales never contain a script subtag? e.g. "zh-Hans-CN".

This regex overall seems very messy. For example, it would allow "en_" or "enUS", or for that matter "foobaren_US!!!". A regex that a bit more clearly matches the language, script, and region subtags and throws away anything after the region might look more like
```
/^([a-z]{2,3})(_[a-z]{4})?(_[a-z]{2}|_[0-9]{3})?(_.*)?$/i
```
Or we might more closely follow the syntax for variants and extensions,
```
/^([a-z]{2,3})(_[a-z]{4})?(_[a-z]{2}|_[0-9]{3})?((?:_[a-z0-9]{5,8}|_[0-9][a-z0-9]{3})+)?((?:_[a-wyz0-9](?:_[a-z0-9]{2,8})+)+)?(?:_x(?:_[a-z0-9]{1,8})+)?$/i
```
(both of these ignore "extlangs", since that part of BCP47 doesn't seem included in the spec for Intl)

But is it even necessary to filter out the variants and extensions? Intl seems like it's supposed to handle or ignore them appropriately.

Reply:
Yeah, I am a regex newbie :\

Here is the full list of WordPress locales today (I should have started there):

https://translate.wordpress.org/

They can be divided into 6 groups:

- 2-letter code (e.g. `af`)
- 3-letter code (e.g. `arq`, that's something I hadn't considered in my regex)
- 2-letter code followed by an underscore and a 2-letter code (e.g. `fr_FR`)
- 3-letter code followed by an underscore and a 2-letter code (e.g. `pap_CW`, also something I hadn't considered)
- 2-letter code followed by an underscore and a 2-letter code and another underscore and a multiple letter code (e.g. `de_DE_formal`)
- `art` followed by an underscore and a multiple letter code (e.g. `art_xpirate` or `art_xemoji`, , also something I hadn't considered)

Although not on the page above, I've also found another option:

- 2-letter code followed by an underscore and a 2-letter code, followed by another underscore and another multiple letter code: `pt_PT_ao90` ([reference](https://make.wordpress.org/polyglots/2017/06/15/pt_pt_ao90-variant-for-pt_pt-locale-request/))

I couldn't find any script subtags, just region subtags and variants.

I think this is what we should aim to support with our regex.

> is it even necessary to filter out the variants and extensions? Intl seems like it's supposed to handle or ignore them appropriately.

That's a good question. I don't know. I wanted to get as close as possible to the canonical version of the locale, but maybe that's not necessary. Do you think we shouldn't bother with it?

Reply:
> Yeah, I am a regex newbie :\

Well, I'm always happy to help. I know way too much regex. (:

> 2-letter code followed by an underscore and a 2-letter code, followed by another underscore and another multiple letter code: pt_PT_ao90

That one's not a valid BCP47 locale. It looks like the "ao90" is supposed to be a variant, but a variant that begins with a letter is supposed to be 5–8 letters/digits long, not 4.

Looks like the official variant for that is "ao1990", according to https://www.iana.org/assignments/language-subtag-registry/language-subtag-registry.

> That's a good question. I don't know. I wanted to get as close as possible to the canonical version of the locale, but maybe that's not necessary. Do you think we shouldn't bother with it?

Absent some belief that there's a browser we want to support that somehow breaks, I'd go for the simpler code and not try to filter it in this way.

On the other hand, since we know there's that invalid "pt_PT_ao90" we might want to add some handling for that sort of thing. Simplest might be something along the lines of
```
locale = cleanLocale( locale );
while ( locale !== '' ) {
    try {
        // Check validity: Intl.NumberFormat will raise an error if the locale is invalid.
        Intl.NumberFormat( locale );
        return locale;
    } catch ( e ) {
    }
    // It was invalid. Strip last component and try again.
    locale = locale.substring( 0, locale.lastIndexOf( '-' ) );
}
```

Reply:
In case you didn't see, I brought up the question at https://make.wordpress.org/polyglots/2017/06/15/pt_pt_ao90-variant-for-pt_pt-locale-request/. The response there was somewhat unsatisfying: they say they don't intend for WordPress's "locales" to map to standard locales at all. OTOH, they also claim that the variant somehow won't be returned anyway, just the "pt_PT".

Reply:
Thank you!

I've made some more changes to try to account for more use-cases; let me know what you think!